ert/data/record/_record.py

```diff
@@ -0,0 +1,224 @@
+
+
+class NumericalRecord(Record):
+    def __init__(
+        self, data: numerical_record_data, index: Optional[RecordIndex] = None
+    ) -> None:
+        if isinstance(data, (set, frozenset, deque, tuple)):
+            data = [val for _, val in enumerate(data)]
+
+        try:
+            self._validate_data(data)
+        except BeartypeException as e:
+            raise RecordValidationError(str(e))
+        self._data = data
+
+        if index is None:
+            index = _build_record_index(data)
+        self._validate_index(data, index)
+        self._index = index
+
+        self._type = self._infer_type(data)
+
+    @beartype
+    def _validate_data(self, data: numerical_record_data) -> None:
+        # beartype does not do deep validation on dicts, so we do non-nested validation
+        if isinstance(data, dict):
+            for key, val in data.items():
+                if not isinstance(key, (str, int)):
+                    raise RecordValidationError(f"unexpected key type {type(key)}")
+                if not isinstance(val, (int, float)):
+                    raise RecordValidationError(f"unexpected value type {type(val)}")
+
+    def _validate_index(
+        self, data: numerical_record_data, index: Optional[RecordIndex] = None
+    ) -> None:
+        norm_record_index = _build_record_index(data)
+        if norm_record_index != index:
+            raise RecordValidationError(
+                f"inconsistent index {norm_record_index} vs {index}"
+            )
+
+    def _infer_type(self, data: numerical_record_data) -> RecordType:
+        if isinstance(data, (list, tuple)):
+            if not data or isinstance(data[0], (int, float)):
+                return RecordType.LIST_FLOAT
+        elif isinstance(data, Mapping):
+            if not data:
+                return RecordType.MAPPING_STR_FLOAT
+            from_ = list(data.keys())[0]
+            if isinstance(from_, (int, float)):
+                return RecordType.MAPPING_INT_FLOAT
+            if isinstance(from_, str):
+                return RecordType.MAPPING_STR_FLOAT
+        raise RecordValidationError(f"unexpected data type {type(data)}")
+
+    @property
+    def index(self) -> RecordIndex:
+        return self._index
+
+    @property
+    def data(self) -> record_data:
+        return self._data
+
+    @property
+    def record_type(self) -> RecordType:
+        return self._type
+
+    def __eq__(self, o: object) -> bool:
+        if isinstance(o, type(self)):
+            return self.__dict__ == o.__dict__
+        return False
+
+
+class RecordCollectionType(str, Enum):
+    NON_UNIFORM = "NON_UNIFORM"
+    UNIFORM = "UNIFORM"
+
+
+_RecordTupleType = Union[Tuple[NumericalRecord, ...], Tuple[BlobRecord, ...]]
+
+
+class RecordCollection:
+    def __init__(
+        self,
+        records: Tuple[Record, ...],
+        ensemble_size: Optional[PositiveInt] = None,
+        collection_type: RecordCollectionType = RecordCollectionType.NON_UNIFORM,
```

Comment:
What's the `NON_UNIFORM` type? I think this type had been defined before.

Reply:
I leave `RecordCollection` largely untouched in this PR, but I move the file. Maybe @verveerpj can answer this?

Reply:
`RecordCollection` was already rewritten in a previous PR,  in an effort to introduce explicit uniform record collections. I.e. collections of identical records, that you do not want to store or transmit multiple times. So, `RecordCollections` gained a `type`: `UNIFORM` or `NON_UNIFORM`.